To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News

Read our **[Qwen3 Guide](https://docs.unsloth.ai/basics/qwen3-how-to-run-and-fine-tune)** and check out our new **[Dynamic 2.0](https://docs.unsloth.ai/basics/unsloth-dynamic-2.0-ggufs)** quants which outperforms other quantization methods!

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [1]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth
else:
    # Do this only in Colab notebooks! Otherwise use pip install unsloth
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft trl==0.15.2 triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf datasets huggingface_hub hf_transfer
    !pip install --no-deps unsloth

### Unsloth

In [2]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 2048 # Choose any! We auto support RoPE Scaling internally!
dtype = None # None for auto detection. Float16 for Tesla T4, V100, Bfloat16 for Ampere+
load_in_4bit = True # Use 4bit quantization to reduce memory usage. Can be False.

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/Meta-Llama-3.1-8B-bnb-4bit",      # Llama-3.1 15 trillion tokens model 2x faster!
    "unsloth/Meta-Llama-3.1-8B-Instruct-bnb-4bit",
    "unsloth/Meta-Llama-3.1-70B-bnb-4bit",
    "unsloth/Meta-Llama-3.1-405B-bnb-4bit",    # We also uploaded 4bit for 405b!
    "unsloth/Mistral-Nemo-Base-2407-bnb-4bit", # New Mistral 12b 2x faster!
    "unsloth/Mistral-Nemo-Instruct-2407-bnb-4bit",
    "unsloth/mistral-7b-v0.3-bnb-4bit",        # Mistral v3 2x faster!
    "unsloth/mistral-7b-instruct-v0.3-bnb-4bit",
    "unsloth/Phi-3.5-mini-instruct",           # Phi-3.5 2x faster!
    "unsloth/Phi-3-medium-4k-instruct",
    "unsloth/gemma-2-9b-bnb-4bit",
    "unsloth/gemma-2-27b-bnb-4bit",            # Gemma 2x faster!
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    # Can select any from the below:
    # "unsloth/Qwen2.5-0.5B", "unsloth/Qwen2.5-1.5B", "unsloth/Qwen2.5-3B"
    # "unsloth/Qwen2.5-14B",  "unsloth/Qwen2.5-32B",  "unsloth/Qwen2.5-72B",
    # And also all Instruct versions and Math. Coding verisons!
    model_name = "Qwen/Qwen2.5-3B-Instruct",
    max_seq_length = max_seq_length,
    dtype = dtype,
    load_in_4bit = load_in_4bit,
    # token = "hf_...", # use one if using gated models like meta-llama/Llama-2-7b-hf
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
==((====))==  Unsloth 2025.4.7: Fast Qwen2 patching. Transformers: 4.51.3.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


model.safetensors:   0%|          | 0.00/2.36G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/271 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/7.36k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/2.78M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/1.67M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/605 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/614 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/11.4M [00:00<?, ?B/s]

We now add LoRA adapters so we only need to update 1 to 10% of all parameters!

In [ ]:
model = FastLanguageModel.get_peft_model(
    model,
    r = 16, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = ["q_proj", "k_proj", "v_proj", "o_proj",
                      "gate_proj", "up_proj", "down_proj",],
    lora_alpha = 16,
    lora_dropout = 0, # Supports any, but = 0 is optimized
    bias = "none",    # Supports any, but = "none" is optimized
    # [NEW] "unsloth" uses 30% less VRAM, fits 2x larger batch sizes!
    use_gradient_checkpointing = "unsloth", # True or "unsloth" for very long context
    random_state = 3407,
    use_rslora = False,  # We support rank stabilized LoRA
    loftq_config = None, # And LoftQ
)

Unsloth 2025.4.7 patched 36 layers with 36 QKV layers, 36 O layers and 36 MLP layers.


<a name="Data"></a>
### Data Prep
We now use the Alpaca dataset from [yahma](https://huggingface.co/datasets/yahma/alpaca-cleaned), which is a filtered version of 52K of the original [Alpaca dataset](https://crfm.stanford.edu/2023/03/13/alpaca.html). You can replace this code section with your own data prep.

**[NOTE]** To train only on completions (ignoring the user's input) read TRL's docs [here](https://huggingface.co/docs/trl/sft_trainer#train-on-completions-only).

**[NOTE]** Remember to add the **EOS_TOKEN** to the tokenized output!! Otherwise you'll get infinite generations!

If you want to use the `llama-3` template for ShareGPT datasets, try our conversational [notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Alpaca.ipynb)

For text completions like novel writing, try this [notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Mistral_(7B)-Text_Completion.ipynb).

In [ ]:
alpaca_prompt = """Below is a medical question followed by a detailed reasoning process. Provide an appropriate response that completes the reasoning.

### Question:
{}

### Reasoning:
{}

### Response:
{}"""



EOS_TOKEN = tokenizer.eos_token # Must add EOS_TOKEN
def formatting_prompts_func(examples):
    questions  = examples["Question"]
    reasonings = examples["Complex_CoT"]
    responses  = examples["Response"]
    texts = []
    for question, reasoning, response in zip(questions, reasonings, responses):
        # Create the prompt with question, reasoning, and response
        text = alpaca_prompt.format(question, reasoning, response) + EOS_TOKEN
        texts.append(text)
    return { "text" : texts, }
pass

from datasets import load_dataset
dataset = load_dataset("FreedomIntelligence/medical-o1-reasoning-SFT", "en", split="train")
# dataset = dataset.select(range(10))
subset_size = int(0.8 * len(dataset))
dataset = dataset.select(range(subset_size))
dataset = dataset.map(formatting_prompts_func, batched = True,)

Map:   0%|          | 0/7881 [00:00<?, ? examples/s]

<a name="Train"></a>
### Train the model
Now let's use Huggingface TRL's `SFTTrainer`! More docs here: [TRL SFT docs](https://huggingface.co/docs/trl/sft_trainer). We do 60 steps to speed things up, but you can set `num_train_epochs=1` for a full run, and turn off `max_steps=None`. We also support TRL's `DPOTrainer`!

In [ ]:
from trl import SFTTrainer
from transformers import TrainingArguments
from unsloth import is_bfloat16_supported

trainer = SFTTrainer(
    model = model,
    tokenizer = tokenizer,
    train_dataset = dataset,
    dataset_text_field = "text",
    max_seq_length = max_seq_length,
    dataset_num_proc = 2,
    packing = False, # Can make training 5x faster for short sequences.
    args = TrainingArguments(
        per_device_train_batch_size = 2,
        gradient_accumulation_steps = 4,
        warmup_steps = 5,
        num_train_epochs = 12, # Set this for 1 full training run.
        # max_steps = 50,
        learning_rate = 2e-4,
        fp16 = not is_bfloat16_supported(),
        bf16 = is_bfloat16_supported(),
        logging_steps = 1,
        optim = "adamw_8bit",
        weight_decay = 0.01,
        lr_scheduler_type = "linear",
        seed = 3407,
        output_dir = "outputs",
        report_to = "none", # Use this for WandB etc
    ),
)

Unsloth: Tokenizing ["text"] (num_proc=2):   0%|          | 0/7881 [00:00<?, ? examples/s]

In [ ]:
# @title Show current memory stats
gpu_stats = torch.cuda.get_device_properties(0)
start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
print(f"GPU = {gpu_stats.name}. Max memory = {max_memory} GB.")
print(f"{start_gpu_memory} GB of memory reserved.")

GPU = Tesla T4. Max memory = 14.741 GB.
3.281 GB of memory reserved.


In [ ]:
trainer_stats = trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 7,881 | Num Epochs = 1 | Total steps = 985
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 4 x 1) = 8
 "-____-"     Trainable parameters = 29,933,568/3,000,000,000 (1.00% trained)


In [ ]:
# @title Show final memory and time stats
used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
used_percentage = round(used_memory / max_memory * 100, 3)
lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
print(f"{trainer_stats.metrics['train_runtime']} seconds used for training.")
print(
    f"{round(trainer_stats.metrics['train_runtime']/60, 2)} minutes used for training."
)
print(f"Peak reserved memory = {used_memory} GB.")
print(f"Peak reserved memory for training = {used_memory_for_lora} GB.")
print(f"Peak reserved memory % of max memory = {used_percentage} %.")
print(f"Peak reserved memory for training % of max memory = {lora_percentage} %.")

462.3942 seconds used for training.
7.71 minutes used for training.
Peak reserved memory = 7.893 GB.
Peak reserved memory for training = 2.129 GB.
Peak reserved memory % of max memory = 53.519 %.
Peak reserved memory for training % of max memory = 14.436 %.


<a name="Inference"></a>
### Inference
Let's run the model! You can change the instruction and input - leave the output blank!

**[NEW] Try 2x faster inference in a free Colab for Llama-3.1 8b Instruct [here](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Unsloth_Studio.ipynb)**

In [ ]:
# alpaca_prompt = Copied from above
FastLanguageModel.for_inference(model) # Enable native 2x faster inference
inputs = tokenizer(
[
    alpaca_prompt.format(
        """A patient presents with hyperacusis, loss of lacrimation and loss of taste sensation in the anterior 2/3rd of the tongue. Oedema extends up to which level of facial nerve -
A. Vertical part
B. Vertical part proximal to nerve to stapedius
C. Vertical part and beyond nerve to stapedius
D. Proximal to geniculate ganglion
""", # instruction
        """Alright, let's think this through. The symptoms at hand start with hyperacusis, which tells us something interesting about the facial nerve. When you get hyperacusis, it usually means that the stapedius muscle isn't working properly, and that muscle is controlled by the nerve to stapedius. So, we know the lesion is around the area where the nerve to stapedius branches off.

Now, let's talk about the loss of lacrimation, or the lack of tears. This usually happens when there's an issue with the greater petrosal nerve. The greater petrosal nerve branches off the facial nerve at the geniculate ganglion, so if lacrimation is affected, the lesion must be at or near this level. Hmm, this starts to narrow things down a bit.

Loss of taste in the anterior 2/3 of the tongue involves the chorda tympani nerve. The chorda tympani is another branch of the facial nerve, but it comes off after the nerve to stapedius. So if taste is affected, this may imply something further along the nerve's path, possibly in the vertical segment.

But wait, if all these symptoms are happening, the lesion can't be too far down the facial nerve. We need to consider how these branches come off sequentially. Hyperacusis suggests involvement with the stapedius, loss of lacrimation indicates an issue at or proximal to the geniculate ganglion, and taste loss could be resultant from swelling or pressure affecting the chorda tympani.

Considering all of this, it makes sense to say that the lesion, or edema, is likely impacting the facial nerve proximal to the geniculate ganglion. This is a spot where you could potentially explain all of these symptom combinations. It impacts the greater petrosal nerve first, causing the loss of tears, and then affects other branches such as those leading to the stapedius and chorda tympani, leading to hyperacusis and taste loss, respectively.

Okay, so all things considered, it seems like the problem is proximal to the geniculate ganglion, where all these branches are close enough together that a singular lesion could affect them simultaneously, leading to the presented symptoms. So, the best answer is definitely D. Proximal to geniculate ganglion.""", # input
        "", # output - leave this blank for generation!
    )
], return_tensors = "pt").to("cuda")

outputs = model.generate(**inputs, max_new_tokens = 256, use_cache = True)
tokenizer.batch_decode(outputs)

["Below is a medical question followed by a detailed reasoning process. Provide an appropriate response that completes the reasoning.\n\n### Question:\nA patient presents with hyperacusis, loss of lacrimation and loss of taste sensation in the anterior 2/3rd of the tongue. Oedema extends up to which level of facial nerve -\nA. Vertical part\nB. Vertical part proximal to nerve to stapedius\nC. Vertical part and beyond nerve to stapedius\nD. Proximal to geniculate ganglion\n\n\n### Reasoning:\nAlright, let's think this through. The symptoms at hand start with hyperacusis, which tells us something interesting about the facial nerve. When you get hyperacusis, it usually means that the stapedius muscle isn't working properly, and that muscle is controlled by the nerve to stapedius. So, we know the lesion is around the area where the nerve to stapedius branches off.\n\nNow, let's talk about the loss of lacrimation, or the lack of tears. This usually happens when there's an issue with the gre

 You can also use a `TextStreamer` for continuous inference - so you can see the generation token by token, instead of waiting the whole time!

In [ ]:
# alpaca_prompt = Copied from above
FastLanguageModel.for_inference(model) # Enable native 2x faster inference
inputs = tokenizer(
[
    alpaca_prompt.format(
        "Continue the fibonnaci sequence.", # instruction
        "1, 1, 2, 3, 5, 8", # input
        "", # output - leave this blank for generation!
    )
], return_tensors = "pt").to("cuda")

from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer)
_ = model.generate(**inputs, streamer = text_streamer, max_new_tokens = 128)

Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
Continue the fibonnaci sequence.

### Input:
1, 1, 2, 3, 5, 8

### Response:
13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811, 


<a name="Save"></a>
### Saving, loading finetuned models
To save the final model as LoRA adapters, either use Huggingface's `push_to_hub` for an online save or `save_pretrained` for a local save.

**[NOTE]** This ONLY saves the LoRA adapters, and not the full model. To save to 16bit or GGUF, scroll down!

In [ ]:
# model.save_pretrained("lora_model")  # Local saving
# tokenizer.save_pretrained("lora_model")
model.push_to_hub("vbhati/lora_model", token = "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") # Online saving
tokenizer.push_to_hub("vbhati/lora_model", token = "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") # Online saving

README.md:   0%|          | 0.00/607 [00:00<?, ?B/s]

  0%|          | 0/1 [00:00<?, ?it/s]

adapter_model.safetensors:   0%|          | 0.00/120M [00:00<?, ?B/s]

Saved model to https://huggingface.co/vbhati/lora_model


  0%|          | 0/1 [00:00<?, ?it/s]

tokenizer.json:   0%|          | 0.00/11.4M [00:00<?, ?B/s]

Now if you want to load the LoRA adapters we just saved for inference, set `False` to `True`:

In [ ]:
if False:
    from unsloth import FastLanguageModel
    model, tokenizer = FastLanguageModel.from_pretrained(
        model_name = "lora_model", # YOUR MODEL YOU USED FOR TRAINING
        max_seq_length = max_seq_length,
        dtype = dtype,
        load_in_4bit = load_in_4bit,
    )
    FastLanguageModel.for_inference(model) # Enable native 2x faster inference

# alpaca_prompt = You MUST copy from above!

inputs = tokenizer(
[
    alpaca_prompt.format(
        "What is a famous tall tower in Paris?", # instruction
        "", # input
        "", # output - leave this blank for generation!
    )
], return_tensors = "pt").to("cuda")

from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer)
_ = model.generate(**inputs, streamer = text_streamer, max_new_tokens = 128)

Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
What is a famous tall tower in Paris?

### Input:


### Response:
One of the most famous tall towers in Paris is the Eiffel Tower. It was built by Gustave Eiffel in 1889 and stands at a height of 324 meters (1,063 feet). It is a symbol of Paris and one of the most recognizable structures in the world.<|endoftext|>


You can also use Hugging Face's `AutoModelForPeftCausalLM`. Only use this if you do not have `unsloth` installed. It can be hopelessly slow, since `4bit` model downloading is not supported, and Unsloth's **inference is 2x faster**.

In [ ]:
if False:
    # I highly do NOT suggest - use Unsloth if possible
    from peft import AutoPeftModelForCausalLM
    from transformers import AutoTokenizer
    model = AutoPeftModelForCausalLM.from_pretrained(
        "lora_model", # YOUR MODEL YOU USED FOR TRAINING
        load_in_4bit = load_in_4bit,
    )
    tokenizer = AutoTokenizer.from_pretrained("lora_model")

### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16 or `merged_4bit` for int4. We also allow `lora` adapters as a fallback. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [ ]:
# Merge to 16bit
if True: model.save_pretrained_merged("model_16bit", tokenizer, save_method = "merged_16bit",)
if True: model.push_to_hub_merged("vbhati/model_16bit", tokenizer, save_method = "merged_16bit", token = "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")

# Merge to 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# Just LoRA adapters
if False: model.save_pretrained_merged("model", tokenizer, save_method = "lora",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "lora", token = "")

Unsloth: You have 1 CPUs. Using `safe_serialization` is 10x slower.
We shall switch to Pytorch saving, which might take 3 minutes and not 30 minutes.
To force `safe_serialization`, set it to `None` instead.
Unsloth: Kaggle/Colab has limited disk space. We need to delete the downloaded
model which will save 4-16GB of disk space, allowing you to save on Kaggle/Colab.
Unsloth: Will remove a cached repo with size 2.4G


Unsloth: Merging 4bit and LoRA weights to 16bit...
Unsloth: Will use up to 3.5 out of 12.67 RAM for saving.
Unsloth: Saving model... This might take 5 minutes ...


100%|██████████| 36/36 [00:01<00:00, 24.32it/s]


Unsloth: Saving tokenizer... Done.
Unsloth: Saving model_16bit/pytorch_model-00001-of-00002.bin...
Unsloth: Saving model_16bit/pytorch_model-00002-of-00002.bin...
Done.


Unsloth: You are pushing to hub, but you passed your HF username = vbhati.
We shall truncate vbhati/model_16bit to model_16bit


Unsloth: Merging 4bit and LoRA weights to 16bit...
Unsloth: Will use up to 3.46 out of 12.67 RAM for saving.
Unsloth: Saving model... This might take 5 minutes ...


100%|██████████| 36/36 [00:00<00:00, 37.91it/s]


Unsloth: Saving tokenizer...

  0%|          | 0/1 [00:00<?, ?it/s]

tokenizer.json:   0%|          | 0.00/11.4M [00:00<?, ?B/s]

 Done.
Unsloth: Saving model_16bit/pytorch_model-00001-of-00002.bin...
Unsloth: Saving model_16bit/pytorch_model-00002-of-00002.bin...


README.md:   0%|          | 0.00/607 [00:00<?, ?B/s]

  0%|          | 0/2 [00:00<?, ?it/s]

pytorch_model-00001-of-00002.bin:   0%|          | 0.00/4.96G [00:00<?, ?B/s]

pytorch_model-00002-of-00002.bin:   0%|          | 0.00/1.21G [00:00<?, ?B/s]

Done.
Saved merged model to https://huggingface.co/vbhati/model_16bit


### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now! We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

[**NEW**] To finetune and auto export to Ollama, try our [Ollama notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)

In [ ]:
# Save to 8bit Q8_0
if False: model.save_pretrained_gguf("model", tokenizer,)
# Remember to go to https://huggingface.co/settings/tokens for a token!
# And change hf to your username!
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "f16", token = "")

# Save to q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "q4_k_m", token = "")

# Save to multiple GGUF options - much faster if you want multiple!
if False:
    model.push_to_hub_gguf(
        "hf/model", # Change hf to your username!
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",],
        token = "", # Get a token at https://huggingface.co/settings/tokens
    )

In [9]:
pip install bitsandbytes

In [3]:
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig

# Load the model and tokenizer from Hugging Face Hub
model_name = "vbhati/model_16bit"
tokenizer = AutoTokenizer.from_pretrained(model_name)

# Configure 4-bit Quantization using bitsandbytes
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,          # Enable 4-bit precision
    quant_type="nf4",           # Normalized float (nf4) for better precision
    compute_dtype="float16"     # Use FP16 for faster computation on GPU
)

# Load model with 4-bit precision
model = AutoModelForCausalLM.from_pretrained(
    model_name,
    quantization_config=bnb_config,
    device_map="auto"  # Automatically place model on GPU if available
)

# Set model to evaluation mode (faster inference)
model.eval()
print("Model loaded in 4-bit precision with bitsandbytes.")

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

Model loaded in 4-bit precision with bitsandbytes.


In [6]:
import pandas as pd
from datasets import load_dataset
from unsloth import FastLanguageModel
import time
import torch
import re

# Testing the model
alpaca_prompt = """Below is a medical question followed by a detailed reasoning process. Provide an appropriate response that completes the reasoning.

### Question:
{}

### Reasoning:
{}

### Response:
{}"""



def formatting_prompts_func(examples):
    questions  = examples["Question"]
    reasonings = examples["Complex_CoT"]
    texts = []
    for question, reasoning in zip(questions, reasonings):
        # Create the prompt with question, reasoning, and response
        text = alpaca_prompt.format(question, reasoning, "")
        texts.append(text)
    return { "text" : texts, }
pass

dataset = load_dataset("FreedomIntelligence/medical-o1-reasoning-SFT", "en", split="train")
# subset_size = int(0.8 * len(dataset))
# subset_size = int(0.8 * len(dataset))
subset_size = int(0.8 * len(dataset))
eval_dataset = dataset.select(range(subset_size, len(dataset)))
eval_dataset = eval_dataset.map(formatting_prompts_func, batched = True,)
FastLanguageModel.for_inference(model)


def extract_response(generated_text):
    response_match = re.search(r"### Response:\s*(.*)", generated_text, re.DOTALL)
    if response_match:
        return response_match.group(1).strip()
    return "[EMPTY RESPONSE]"

def get_prediction(model, tokenizer, eval_dataset):
  predictions = []
  actual = []
  question = []
  chain_of_thoughts = []
  total_examples = len(eval_dataset)
  start_time = time.time()
  for index, example in enumerate(eval_dataset):
    input_text = example["text"]
    inputs = tokenizer(input_text, return_tensors="pt").to("cuda")
    with torch.no_grad():
      output = model.generate(**inputs, max_new_tokens = 256, use_cache = True)
    get_op = tokenizer.decode(output[0], skip_special_tokens=True)
    # print(get_op)
    model_pred = extract_response(get_op)
    # print(model_pred)
    actual.append(example["Response"])
    predictions.append(model_pred)
    question.append(example["Question"])
    chain_of_thoughts.append(example["Complex_CoT"])
    elapsed_time = time.time() - start_time
    progress = (index + 1) / total_examples
    estimated_total_time = elapsed_time / progress
    remaining_time = estimated_total_time - elapsed_time

    print(f"\rProcessing: {index + 1}/{total_examples} ({progress * 100:.2f}%) - Time Elapsed: {elapsed_time:.2f}s - Estimated Time Remaining: {remaining_time:.2f}s", end="")
  df_predictions = pd.DataFrame({
      'question': question,
      'chain_of_thoughts': chain_of_thoughts,
      'actual': actual,
      'model_output': predictions
      })
  return df_predictions

predictions_df = get_prediction(model, tokenizer, eval_dataset)
predictions_df.to_csv("model_predictions.csv", index=False)

predictions_df.head()

<ipython-input-6-1d0457b3efcc>:3: UserWarning: WARNING: Unsloth should be imported before transformers to ensure all optimizations are applied. Your code may run slower or encounter memory issues without these optimizations.

Please restructure your imports with 'import unsloth' at the top of your file.
  from unsloth import FastLanguageModel


🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
Processing: 33/100 (33.00%) - Time Elapsed: 575.68s - Estimated Time Remaining: 1168.81s

KeyboardInterrupt: 

In [ ]:
import pandas as pd
from google import genai
from google.genai import types
import re
import time

# Initialize the GenAI Client with your API Key
client = genai.Client(api_key="API-KEY")  # Replace with your API key

# Define the LLM Judge Prompt
judge_prompt = """You are an expert judge evaluating the accuracy of a model-generated response. Based on the input, the actual answer, and the model's answer, please rate the model's response on a scale of 1 to 5, with 5 being perfect accuracy and 1 being completely incorrect.

### Question:
{}

### Actual Answer:
{}

### Model-Generated Response:
{}

### Score (1-5):
Score:"""

# Load your predictions CSV
predictions_df = pd.read_csv("model_predictions_PubMedQA.csv")

# Function to generate score using Gemini 2.0 Flash (Google GenAI API)
def generate_gemini_score(question, actual_text, model_text, max_retries=3):
    # Create the evaluation prompt
    input_text = judge_prompt.format(question, actual_text, model_text)

    for attempt in range(max_retries):
        try:
            # Send request to Gemini 2.0 Flash
            response = client.models.generate_content(
                model="gemini-2.0-flash",
                contents=[input_text],
                config=types.GenerateContentConfig(
                    max_output_tokens=100,  # Adjust as needed
                    temperature=0.1)
            )

            # Extract the generated text directly
            generated_text = response.text.strip()
            print("\nGemini 2.0 Flash Evaluation Output:\n", generated_text)

            # Improved regex to capture any numeric score (1-5)
            score_match = re.search(r"\b([1-5])\b", generated_text)
            if score_match:
                score = int(score_match.group(1))
                print(f"Extracted Score: {score}")
                return score

            print("Score not found. Defaulting to 0.")
            return 0
        except Exception as e:
            print(f"Error on attempt {attempt + 1}: {e}")
            if attempt < max_retries - 1:
                print("Retrying...")
                time.sleep(3)  # Small delay before retrying
            else:
                print("Max retries reached. Defaulting to 0.")
                return 0

# Apply Gemini 2.0 Flash as Judge
predictions_df["llm_score"] = predictions_df.apply(
    lambda row: generate_gemini_score(row["input"], row["actual"], row["model_output"]), axis=1
)

# Save the updated CSV with LLM Scores
predictions_df.to_csv("model_predictions_with_gemini_scores.csv", index=False)
print("\nGemini 2.0 Flash Scored Predictions saved to model_predictions_with_gemini_scores.csv")

# Display the results
print(predictions_df.head())



Gemini 2.0 Flash Evaluation Output:
 4
Extracted Score: 4

Gemini 2.0 Flash Evaluation Output:
 5
Extracted Score: 5

Gemini 2.0 Flash Evaluation Output:
 2
### Explanation:
The model's response is inaccurate and contains information not found in the source document. The actual answer states that preoperative imaging is only accurate in 75% of cases, while the model claims a 100% interobserver agreement, which is contradictory and not supported by the provided context. Additionally, the model introduces concepts like Simpson grade 1 resection and specific sensitivity/specificity values (89% and 52%) that are not mentioned in the original
Extracted Score: 2

Gemini 2.0 Flash Evaluation Output:
 1
Extracted Score: 1

Gemini 2.0 Flash Evaluation Output:
 3
Extracted Score: 3

Gemini 2.0 Flash Evaluation Output:
 1
### Rationale:
The model-generated response is nonsensical and does not accurately address the question. It contains grammatical errors, repetitive phrases, and unclear stateme

In [4]:
# Lets use different datasets
import pandas as pd
from datasets import load_dataset
# from unsloth import FastLanguageModel
import time
import torch
import re

# Testing the model
alpaca_prompt = """Below is a medical question with detailed Instructions. Provide an appropriate response to the Question that complies with instructions.

### Instruction:
{}

### Input:
{}

### Output:
{}"""



def formatting_prompts_func(examples):
    instructions = examples["instruction"]
    inputs = examples["input"]
    outputs = examples["output"]
    texts = []
    for instruction, input_text, output in zip(instructions, inputs, outputs):
        # Create the prompt with question, reasoning, and response
        text = alpaca_prompt.format(instruction, input_text, "")
        texts.append(text)
    return { "text" : texts, }
pass

dataset = load_dataset("llamafactory/PubMedQA", split="train")
# subset_size = int(0.8 * len(dataset))
subset_size = int(0.8 * len(dataset))
# subset_size = len(dataset) - 100
eval_dataset = dataset.select(range(subset_size, len(dataset)))
eval_dataset = eval_dataset.map(formatting_prompts_func, batched = True,)
FastLanguageModel.for_inference(model)


def extract_response(generated_text):
    response_match = re.search(r"### Output:\s*(.*)", generated_text, re.DOTALL)
    if response_match:
        return response_match.group(1).strip()
    return "[EMPTY RESPONSE]"

def get_prediction(model, tokenizer, eval_dataset):
  predictions = []
  actual = []
  input_texts = []
  instructions = []

  total_examples = len(eval_dataset)
  start_time = time.time()

  for index, example in enumerate(eval_dataset):
    input_text = example["text"]
    inputs = tokenizer(input_text, return_tensors="pt").to("cuda")
    with torch.no_grad():
      output = model.generate(**inputs, max_new_tokens = 256, use_cache = True)
    get_op = tokenizer.decode(output[0], skip_special_tokens=True)

    # print(get_op)
    model_pred = extract_response(get_op)
    # print(model_pred)

    actual.append(example["output"])
    predictions.append(model_pred)
    input_texts.append(example["input"])
    instructions.append(example["instruction"])

    elapsed_time = time.time() - start_time
    progress = (index + 1) / total_examples
    estimated_total_time = elapsed_time / progress
    remaining_time = estimated_total_time - elapsed_time

    print(f"\rProcessing: {index + 1}/{total_examples} ({progress * 100:.2f}%) - Time Elapsed: {elapsed_time:.2f}s - Estimated Time Remaining: {remaining_time:.2f}s", end="")
  df_predictions = pd.DataFrame({
      'input': input_texts,
      'instructions': instructions,
      'actual': actual,
      'model_output': predictions
      })
  return df_predictions

predictions_df = get_prediction(model, tokenizer, eval_dataset)
predictions_df.to_csv("model_predictions_PubMedQA.csv", index=False)

predictions_df.head()

/usr/local/lib/python3.11/dist-packages/bitsandbytes/nn/modules.py:451: UserWarning: Input type into Linear4bit is torch.float16, but bnb_4bit_compute_dtype=torch.float32 (default). This will lead to slow inference or training speed.
  warnings.warn(


Processing: 100/100 (100.00%) - Time Elapsed: 1614.56s - Estimated Time Remaining: 0.00s                                               input  \
0  Question: Total knee arthroplasty using patien...   
1  Question: Is Canine Eruption Velocity Affected...   
2  Question: Can Preoperative Imaging Predict Tum...   
3  Question: Care-Dependency in Parkinson's Disea...   
4  Question: Is There an Enabling Environment for...   

                                        instructions  \
0  Answer the question based on the following con...   
1  Answer the question based on the following con...   
2  Answer the question based on the following con...   
3  Answer the question based on the following con...   
4  Answer the question based on the following con...   

                                              actual  \
0  This study suggests that there is a minimal le...   
1  The use of allograft bone to augment bone graf...   
2  Preoperative imaging of clinoid region meningi...   
3  ADL impairm

In [7]:
import pandas as pd
from datasets import load_dataset
import time
import torch
import re
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig


# Define the Alpaca-style Prompt
alpaca_prompt = """Below is a medical question with detailed instructions. Provide an appropriate response to the Question that complies with instructions.

### Instruction:
{}

### Input:
{}

### Output:
{}"""

# Load the Dataset (bigbio/med_qa)
dataset = load_dataset("bigbio/med_qa", split="train")
subset_size = int(0.8 * len(dataset))  # Last 20% rows

# Select the last 20% rows
eval_dataset = dataset.select(range(subset_size, len(dataset)))
print(f"Loaded Dataset with {len(dataset)} examples.")

# Formatting Function for Prompts
def formatting_prompts_func(examples):
    instructions = examples["instruction"] if "instruction" in examples else [""] * len(examples["question"])
    inputs = examples["question"]
    outputs = examples.get("answer", [""] * len(inputs))  # Default empty output if not present
    texts = []

    for instruction, input_text, output in zip(instructions, inputs, outputs):
        text = alpaca_prompt.format(instruction, input_text, "")
        texts.append(text)

    return { "text" : texts }

# Prepare Evaluation Dataset
eval_dataset = dataset.map(formatting_prompts_func, batched=True)

# Define the Response Extraction Function
def extract_response(generated_text):
    response_match = re.search(r"### Output:\s*(.*)", generated_text, re.DOTALL)
    if response_match:
        return response_match.group(1).strip()
    return "[EMPTY RESPONSE]"

# Prediction Function
def get_prediction(model, tokenizer, eval_dataset):
    predictions = []
    actual = []
    input_texts = []
    instructions = []

    total_examples = len(eval_dataset)
    start_time = time.time()

    for index, example in enumerate(eval_dataset):
        input_text = example["text"]
        inputs = tokenizer(input_text, return_tensors="pt", padding=True, truncation=True).to("cuda")

        with torch.no_grad():
            output = model.generate(**inputs, max_new_tokens=256, use_cache=True)

        # Decode and Extract Response
        get_op = tokenizer.decode(output[0], skip_special_tokens=True)
        model_pred = extract_response(get_op)

        # Collect Results
        actual.append(example.get("answer", "[NO ANSWER]"))  # Use the correct field from dataset
        predictions.append(model_pred)
        input_texts.append(example["question"])
        instructions.append(example.get("instruction", "[NO INSTRUCTION]"))

        # Progress Tracking
        elapsed_time = time.time() - start_time
        progress = (index + 1) / total_examples
        estimated_total_time = elapsed_time / progress
        remaining_time = estimated_total_time - elapsed_time

        print(f"\rProcessing: {index + 1}/{total_examples} ({progress * 100:.2f}%) - Time Elapsed: {elapsed_time:.2f}s - Estimated Time Remaining: {remaining_time:.2f}s", end="")

    # Save Results to DataFrame
    df_predictions = pd.DataFrame({
        'input': input_texts,
        'instructions': instructions,
        'actual': actual,
        'model_output': predictions
    })
    return df_predictions

# Generate Predictions
predictions_df = get_prediction(model, tokenizer, eval_dataset)
predictions_df.to_csv("model_predictions_med_qa.csv", index=False)

# Display the Results
print("\nSample Predictions:")
print(predictions_df.head())


README.md:   0%|          | 0.00/1.44k [00:00<?, ?B/s]

med_qa.py:   0%|          | 0.00/11.1k [00:00<?, ?B/s]

bigbiohub.py:   0%|          | 0.00/19.3k [00:00<?, ?B/s]

The repository for bigbio/med_qa contains custom code which must be executed to correctly load the dataset. You can inspect the repository content at https://hf.co/datasets/bigbio/med_qa.
You can avoid this prompt in future by passing the argument `trust_remote_code=True`.

Do you wish to run the custom code? [y/N] y


data_clean.zip:   0%|          | 0.00/132M [00:00<?, ?B/s]

Generating train split: 0 examples [00:00, ? examples/s]

/root/.cache/huggingface/datasets/downloads/extracted/454fe0fdcb22e71e2762c35bcbe1c9a23f7b0154ade8598b60ee0dc6db7d537a/data_clean/questions/US/train.jsonl


Generating test split: 0 examples [00:00, ? examples/s]

/root/.cache/huggingface/datasets/downloads/extracted/454fe0fdcb22e71e2762c35bcbe1c9a23f7b0154ade8598b60ee0dc6db7d537a/data_clean/questions/US/test.jsonl


Generating validation split: 0 examples [00:00, ? examples/s]

/root/.cache/huggingface/datasets/downloads/extracted/454fe0fdcb22e71e2762c35bcbe1c9a23f7b0154ade8598b60ee0dc6db7d537a/data_clean/questions/US/dev.jsonl
Loaded Dataset with 100 examples.


Map:   0%|          | 0/100 [00:00<?, ? examples/s]

Processing: 100/100 (100.00%) - Time Elapsed: 1676.79s - Estimated Time Remaining: 0.00s
Sample Predictions:
                                               input      instructions  \
0  A 23-year-old pregnant woman at 22 weeks gesta...  [NO INSTRUCTION]   
1  A 3-month-old baby died suddenly at night whil...  [NO INSTRUCTION]   
2  A mother brings her 3-week-old infant to the p...  [NO INSTRUCTION]   
3  A pulmonary autopsy specimen from a 58-year-ol...  [NO INSTRUCTION]   
4  A 20-year-old woman presents with menorrhagia ...  [NO INSTRUCTION]   

                                              actual  \
0                                     Nitrofurantoin   
1  Placing the infant in a supine position on a f...   
2       Abnormal migration of ventral pancreatic bud   
3                                    Thromboembolism   
4                             Von Willebrand disease   

                                        model_output  
0  ###Instruction:\nIn this case, the patient is ...  

In [11]:
pip install -U bitsandbytes

In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig
import math
# Load the model and tokenizer from Hugging Face Hub
model_name = "vbhati/model_16bit"
tokenizer = AutoTokenizer.from_pretrained(model_name)

# Configure 4-bit Quantization using bitsandbytes
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,          # Enable 4-bit precision
    quant_type="nf4",           # Normalized float (nf4) for better precision
    compute_dtype="float16"     # Use FP16 for faster computation on GPU
)

# Load model with 4-bit precision
model = AutoModelForCausalLM.from_pretrained(
    model_name,
    quantization_config=bnb_config,
    device_map="auto"  # Automatically place model on GPU if available
)

# Set model to evaluation mode (faster inference)
model.eval()
print("Model loaded in 4-bit precision with bitsandbytes.")

# List of CSVs to Process
dataframes = ["model_predictions_base.csv", "model_predictions_med_qa.csv", "model_predictions_PubMedQA.csv"]

# Function to Calculate Perplexity
def calculate_perplexity(model, tokenizer, text):
    inputs = tokenizer(text, return_tensors="pt", truncation=True).to("cuda")
    input_ids = inputs.input_ids

    with torch.no_grad():
        outputs = model(input_ids, labels=input_ids)
        log_likelihood = outputs.loss.item()  # Cross-Entropy Loss

    # Calculate Perplexity
    perplexity = math.exp(log_likelihood)
    return perplexity

# Process Each CSV File
for csv_file in dataframes:
    print(f"\nProcessing: {csv_file}")

    # Load the DataFrame with Model Responses
    predictions_df = pd.read_csv(csv_file)
    print("\nLoaded DataFrame:")
    print(predictions_df.head())

    # Calculate Perplexity for Each Response
    perplexity_scores = []

    for response in predictions_df['model_output']:
        perplexity = calculate_perplexity(model, tokenizer, response)
        perplexity_scores.append(perplexity)

    # Add Perplexity Scores to DataFrame
    predictions_df['perplexity'] = perplexity_scores

    # Save the Updated DataFrame
    output_file = csv_file.replace(".csv", "_with_perplexity.csv")
    predictions_df.to_csv(output_file, index=False)
    print(f"\nPerplexity Scores added and saved to {output_file}")

    # Display the Updated Results
    print(predictions_df.head())


Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
